cvat/apps/engine/views.py

```diff
@@ -886,19 +1057,48 @@ def upload_finished(self, request):
             task_data = self._object.data
             serializer = DataSerializer(task_data, data=request.data)
             serializer.is_valid(raise_exception=True)
-            data = dict(serializer.validated_data.items())
-            uploaded_files = task_data.get_uploaded_files()
-            uploaded_files.extend(data.get('client_files'))
-            serializer.validated_data.update({'client_files': uploaded_files})
+            data = serializer.validated_data
 
+            # Merge file lists from TUS and simple/multi uploading
+            uploaded_files = task_data.get_uploaded_files()
+            if data.get('client_files'):
+                if not uploaded_files:
+                    # This request is the only request in the uploading.
+                    # We need to remember file ordering as it may be lost,
+                    # e.g. on a server crash. The input file order could be
+                    # restored from this variable otherwise.
+                    self._init_tus_custom_ordering([f['file'].name for f in data['client_files']])
+
+                response = self.append_files(request)
+                if not 200 <= response.status_code < 300:
+                    return response
+                uploaded_files = task_data.get_uploaded_files()
+
+            data.update({'client_files': uploaded_files})
+
+            # Refresh the db value with the updated file list and other request parameters
             db_data = serializer.save()
             self._object.data = db_data
             self._object.save()
-            data = {k: v for k, v in serializer.data.items()}
 
+            # Create a temporary copy of the parameters we will try to create the task with
+            # and sync on success
+            data = copy(serializer.data)
             data['use_zip_chunks'] = serializer.validated_data['use_zip_chunks']
             data['use_cache'] = serializer.validated_data['use_cache']
             data['copy_data'] = serializer.validated_data['copy_data']
+
+            if data['client_files']:
+                # Django will sort client_files by filename automatically, which conflicts
+                # with the "predefined" sorting method. We need to restore the upload file
+                # order. This is not needed for other *_files fields, because they are
+                # not being sorted.
+                try:
+                    if data['sorting_method'] == models.SortingMethod.PREDEFINED:
+                        data['client_files'] = self._sort_uploaded_files(data['client_files'])
```

Comment:
Should sorting also be applied to `server_files`?

Since we have issues with file order when creating a task with files on a connected file share while using a POST request to `/api/tasks/{id}/data` API, where files are passed as `server_files`.

Reply:
Hi, thank you for the comment. In all cases except client_files with TUS uploading, we pass the ordered list of files in the request. Could you please describe the use case and the problem in more details?

Reply:
Hi! We use REST API of CVAT to programmatically create CVAT tasks. Our files are mostly stored on a mounted file share, so we pass them as `server_files`. 

The order of files of the resulting CVAT task does not match the original order in a list of  `server_files` passed to a POST request to `/api/tasks/{id}/data`. 

I launched CVAT from `zm/fix-predefined-sorting` to test if it helps in our case, and unfortunately the order was not maintained.

I added a debug print directly before line 1091 of `views.py` (`if data['client_files']:`) and it shows that the order of `server_files` is violated somewhere before this point.

Reply:
Thank you for reporting, I'll investigate this problem. Could you please attach the request you're sending so I could reproduce it?

Reply:
I now added one more debug print of `request.data` after line 1056 of `views.py` (`elif self.action == 'data':`) and the order of `server_files` was correct at this point, so something changes it between lines 1056 and 1091, e.g. deserialization.

Reply:
>Thank you for reporting, I'll investigate this problem. Could you please attach the request you're sending so I could reproduce it?

Sure, here is the body of a POST request to `api/tasks/{id}/data`:
https://gist.github.com/OutSorcerer/c485498a06a6d3ac44df42de75cdc5fe

It is quite long, but I was not able to reproduce it with small amount of files. I replaced the customer name in file names with "customer-name" string, this is the only difference.

Here is an example of sorting at line 1091 at `data['server_files']` (it displays 2 characters before the dot in a file name, originally files where sorted in reverse order by these values, but CVAT sorts them mostly in ascending):

https://gist.github.com/OutSorcerer/a1351628ef57c5fb7dc9d8262093271a

Reply:
Ok, thanks. One possible reason is DB saving/loading (in `serializer.save()`) - I've found that for some reason Django sorts `client_files`, and this may be true for other file lists as well.

Reply:
Looks like it is because PostgreSQL [does not maintain insertion order in general](https://dba.stackexchange.com/questions/95822/does-postgres-preserve-insertion-order-of-records). Django has [order-with-respect-to](https://docs.djangoproject.com/en/4.1/ref/models/options/#order-with-respect-to) for cases like this.

Reply:
I've added code to restore DB insertion order in the DB fetching results. Thank you for your help.